### DISCUSSION TITLE
OTA update fetched by EAS dev client but not installed on reload

### DISCUSSION THREAD

**Discussion by lukemillur** [expo/expo#17230] [Archived: Questions]
**Summary** 

This issue is occurring on IOS. After following along with the [installing app variants on the same device](https://docs.expo.dev/build-reference/variants/) documentation, I have two versions of my app installed on my device. The first is a staging deployment installed via TestFlight with a bundle id of `com.projectOwner.projectName` and the second is a development deployment created via EAS build with a bundle id of `com.projectOwner.projectName.dev`. The problem occurs on the development client.

When an update is published the development client successfully recognizes/retrieves the update, however, on app reload (in both manual and automatic update configuration) the changes have not been installed. Here is the return value `Updates.fetchUpdateAsync` captured through Sentry:

![image](https://user-images.githubusercontent.com/32378674/165596088-f4ac4e57-ba52-4551-a01f-5801d8c2a06c.png)

In the past I have been able to publish OTA updates to this project via `expo publish` without specifying a release channel. Unfortunately I can't upgrade to a EAS production plan at the moment, so I have to sort this out with classic updates.

I've included all the information that I feel is relevant to the problem but I'm sure I've missed something, please let me know if there is anything else that may aid in solving the problem. I've been at this for days now, any and all potential paths forward would be greatly appreciated. 

Thanks in advance!

---

**expo-env-info output** 

![image](https://user-images.githubusercontent.com/32378674/165595446-7c611fc2-ea34-4144-8b45-d5ffb3de8fd5.png)

**eas.json**

![image](https://user-images.githubusercontent.com/32378674/165595499-2fbfc591-1af9-49be-a050-195ebaf381a9.png)

**app.config.js**

![image](https://user-images.githubusercontent.com/32378674/165595574-d59b3b7f-d38e-4272-b5e1-cc16b8a6a280.png)

**info.plist**

![image](https://user-images.githubusercontent.com/32378674/165595715-51e65262-f620-4784-8bc6-0e3c79738f56.png)

**Expo.plist**

![image](https://user-images.githubusercontent.com/32378674/165595965-73545c64-3402-4a41-b6d2-e4217e0027ad.png)

**Relevant package info**

- Expo updates version: 0.11.7
- Expo constants version: 13.0.1
- Expo dev client version: 0.8.5

**Managed or bare workflow?**

- bare

**What I have tried**
- Automatic OTA updates
- Manual OTA updates 
- Uninstalling the TestFlight app so that only the development client is installed.
- I have tried every possible configuration of variables in relevant files (`app.config.js`, `eas.json`, `Expo.plist`, `Info.plist`) to no avail.
- Configuration of `expo-splash-screen` to try and provide a highly compatible interim step while the update is fetched and the app is reloaded via `expo-updates`. 

**Miscellaneous Observations**
- The native app appears to have no manifest initially.
- Programmatically changing the name of the app via `app.config.js` renames all previously created builds in the expo dev console.
- The transform cache is reset during the `Bundle React Native code and Images` build phase.
- `main.jsbundle` is missing in Xcode (red text)
- I've been building my dev clients locally to cut down on build times. I am always greeted with this error while running FastLane but at no point has it ever caused the build to fail. I've tried deleting my `DerivedData` but it doesn't seem to have an effect either way. I'm including the log on the off chance it may be relevant to the problem.
![image](https://user-images.githubusercontent.com/32378674/165596007-201f22c5-a129-4662-8a48-abda697e011e.png)

**Comment by Emex4gman** [comment]
It still sucks as there is no proper documentation for building and installing different variants of the same app view eas or even locally. this [Guild](https://docs.expo.dev/build-reference/variants/) doesn't help.

**Reply by lukemillur** [reply]
I actually found the guide you linked to be fairly helpful. So long as you step through the process closely the guide leads you to a point at which you can have multiple app variants installed on the same device. I just wish there was _any_ sort of documentation or precedence for installing OTA updates to app variants. 

What's the point of an app variant for your staging deployment if you can't efficiently push your changes?

**Reply by Emex4gman** [reply]
Sadly there were some missing steps on Xcode, that made the eas build command to fail, when working with varients, but we were able to resolve this. 
To be honest, id rather just stick with one variant. because I found out that when I have both apps installed, and I am working locally, scanning the local QR code opens the wrong variant, I have to uninstall one to use my main dev client build. This might not be the case for other people. Was only able to get the variant setup to work for ios, wasn't able to get it to work for android and it hasn't affected our android delopyment and testing in any way.

This just my two cents though, I am pretty sure it's working for some people. I hope you are able to resolve yours.

**Comment by MorenoMdz** [comment]
> **Summary**
> 
> This issue is occurring on IOS. After following along with the [installing app variants on the same device](https://docs.expo.dev/build-reference/variants/) documentation, I have two versions of my app installed on my device. The first is a staging deployment installed via TestFlight with a bundle id of `com.projectOwner.projectName` and the second is a development deployment created via EAS build with a bundle id of `com.projectOwner.projectName.dev`. The problem occurs on the development client.
> 
> When an update is published the development client successfully recognizes/retrieves the update, however, on app reload (in both manual and automatic update configuration) the changes have not been installed. Here is the return value `Updates.fetchUpdateAsync` captured through Sentry:
> 
> ![image](https://user-images.githubusercontent.com/32378674/165596088-f4ac4e57-ba52-4551-a01f-5801d8c2a06c.png)
> 
> In the past I have been able to publish OTA updates to this project via `expo publish` without specifying a release channel. Unfortunately I can't upgrade to a EAS production plan at the moment, so I have to sort this out with classic updates.
The installed version that is not applying the OTA update is the TestFlight one?

**Reply by lukemillur** [reply]
Correct, I have been unable to install OTA updates on the TestFlight release, a standalone client (non app variant), or an app variant (differing bundle ID).

**Comment by lukemillur** [comment]
Update: I have abandoned expo-updates in favor of eas-updates with the hope that eas-updates would solve the problem. 

Unfortunately it has not.

 I'm still running into the same issue where my app variant can fetch the eas-update but when installing the update itself (both automatic & manual) no changes take effect. 

I've managed to uncover some semi-useful logs from the IOS device console where it appears as though the issue lies in the attempted install location of the update. With that being said I haven't really made any progress following down this path as the device console logs are full of never ending rabbit holes.

Looking for any and all insight that may be of use here.

Thanks in advance!

**Reply by GtechHardik** [reply]
I am also facing the same issue, but would like to point that  eas update uses expo-updates to function, so expo update is not abandoned as mentioned by lakemillur

[Update is a hosted service that serves updates for projects using the expo-updates library. refer below mentioned official link 
](https://docs.expo.dev/eas-update/introduction/)


however i am also facing same issue while migrated from classic build system, eas is not updating ota client side

waiting for some solution

Thank you all

**Comment by GringoDotDev** [comment]
Just curious - if you force close the app and open it again, does the update then load?

I'm running into an issue right now with EAS update where OTA updates are not loaded on app start (e.g. a fresh install), but do indeed load after force closing the app and opening it again.

In other words, first run experience is broken for users. For now I have to abandon OTA updates until it gets figured out.

**Reply by GringoDotDev** [reply]
Well, I should have known that as soon as I posted in public I would figure it out! Writing this up for future travelers.

First, read the writeup here. Skip to the practical overview at the bottom. https://docs.expo.dev/eas-update/how-eas-update-works/

There is a field called `expo.updates.fallbackToCacheTimeout` that controls how long the app will wait before falling back to its cached version. In the case of first run, that will be the js bundled with the app store build.

*It defaults to zero!*

Change that to something more reasonable and the app will wait to load your most recent js before loading.

**Reply by GtechHardik** [reply]
Thanks jseitel, for replying

i kept expo.updates.fallbackToCacheTimeout to zero only

after force stopping / killing app several times app does not update.  

My app is currently not on play store, i am currently generating apk and distributing to internal group for real time experience.

With classic build, i was able to get successfully OTA after every build and client app automatically got updated every time.

But, after switching to eas, neither eas build OTA updates client app, nor eas update. And now i am unable to switch back to classic system

I am fairly new to this system , kindly help 

Kindly let me know 
1) difference between eas build and eas updates.  
2) how an update is attached to build. [ i have tried different channels, branch, update etc] 
3) successfull update has option of preview but, how to connect it to build and publish to build / or say client apps
4) i have doubt that eas update does not work with managed work flow

I am stuck with this issue since last one week

Thanks in advance

**Reply by Emex4gman** [reply]
I have a question, for your Internal build(where developmentClient is set to false }, what eas profile do you use and what command do you run when your run the eas update?

**Reply by GtechHardik** [reply]
Thanks for replying

I tried with eas profile 
1) production 
2) shripad (custom name)

'Default' as  Channel

I tried eas build --profile production
and eas build --profile shripad (latest one)

Both with developmentClient false in eas.json
After build I tried eas update with default channel

All above executes successfully, but does not OTA update,

Every time I also increase verson and version code

**Reply by Emex4gman** [reply]
have you tried to run 
`eas update --branch shripad --platform all --message "shripad update"`

**Reply by Emex4gman** [reply]
For the expo.updates.fallbackToCacheTimeout, best not to use 0
update the value to like 5000, which is 5 seconds 

run this command to update the native folders
`expo prebuild`

then run
`eas build --profile production` or `eas build --profile shripad`

Install the new builds and test the updates.

**Reply by GtechHardik** [reply]
Ok, I have not run expo prebuild,

I will run, and update what happened 

Please also let me know, How builds and update connected,  
I mean how an update knows which build to update?

Thanks million to reply

**Reply by Emex4gman** [reply]
TL;DR
An update is tied to the `channel` or `releaseChannel` in a build profile in your eas.json file and you use `--branch` to specify the channel you want to ship the update to in the `eas update` command
 
For `eas update`, the --branch is tied to the channel or releaseChannel in a build profile (I prefer to use channel), and this is the approach I use. you can point a build to an update, but it becomes hard to track. 
So for example. if you have a build profile called `prod-qa` in you `eas.json` file like this, 
```
{
    ...

    "prod-qa": {
        "channel": "prod-qa",
        "developmentClient": false,
        "distribution": "internal"
         ...
    }

   ...
}
```

You can send an update to this profile build, by running the comment

`eas update --branch prod-qa --platform all --message "prod-qa update"`

the `--branch` will specify the channel to send the update to. 

you can apply this same to your production build profile and run 

`eas update --branch production --platform all --message "production update"`

**Reply by GtechHardik** [reply]
Thank you for providing clarity on branch and channel

tried with 


`expo prebuild`

then, created freash build (just to ensure testing with freash build)

`eas update --branch production --platform android  --message "update..."`  

then

`eas update --branch production --platform all --message "production update`

but still it is not OTA updating.

I think there is issue with runtime version, 
any idea how to set runtimeVersion in app.json and AndroidMainfest.xml

`eas update:configure`  modified  app.json and AndroidMainfest.xml with version : '1.0.0'

if i configure with runtimeVersion '1.0.0' it is build and update published (OTA not happpend), and also `expo start `expo go is not opening and showing incompatible version error
if i configure runtimeVersion : {policy : "sdkVersion"} in app.json, expo go works ok, but than i dont know what to mention in AndroidMainfest.xml
if i remove AndroidMainfest.xml runtimeVersion value, `eas update ` gives error that android version is missing

so, what we should mention in runtimeVersion , our own version numbering or expo sdk version. doc

And the ultimate issue remains same that OTA update is not happening

I gone through documentation aswell but, it didn't help reolving issue

**Reply by Emex4gman** [reply]
Can you send a screenshot of the build page for the production on expo.dev

**Reply by GtechHardik** [reply]
Thank you very very much [Emex4gman](https://github.com/Emex4gman), your explanation and suggestions have helped a lot

finally i am able to OTA update. resolved the issue.

ok, so what i was missing was 
1) even you do not have any native code you have to do `eas prebuild` 
2) runtimeVersion is the expo sdk version required to set in AndroidMainfest.xml and app.json.  with `eas update:configure` it will set to default 'sdkVersion' in app.json and 1.0.0 in AndroidMainfest.xml, which was misleading to me.  finally i placed 45.0.0 as my runtimeVersion in both app.json and AndroidMainfest.xml, it worked like charm.  

What i learned is  
a) we cannot put 1.0 like single decimal version, it will not give error but will not work aswell,. it should be like 1.0.0 etc. there is eas policy for the same, but it does not mention this.
b) put actual expo sdkversion  in app.json and AndroidMainfest.xml, and change values set by `eas update:configure` . 
d) if you put runtimeVersion as policy : 'sdkVersion'  in app.json and AndroidMainfest.xml, with 1.0.0 etc. it will not work. what i did is set my runtimeVersion 45.0.0 all over and that fine with me. (i run on exposdk 45)
c) i dont know about putting policy : 'sdkVersion' in AndroidMainfest.xml i haven't tried.
d) app.json goes with each time you build so, changing it after build while updating i.e. `eas update` will not work. new build is required after every change in app.json

**Comment by selimjouan** [comment]
I use a **managed workflow** (**iOS**, with **multiple variants**) but I have the exact same problem, and tried the same things.
I even tried downgrading from SDK 47 to 46. No results.

However I got this native error when updating (automatically and manually):

```
EXUpdatesAppController _emergencyLaunchWithFatalError
Fatal Exception: RCTFatalException: Unhandled JS Exception: Error: token is not a valid string
```

Does somebody know what it means?

**Reply by selimjouan** [reply]
I managed to resolve it.
During `eas update`, environment variables (set during build) are set to null. You have to set them again in the command line when updating:
`API_URL1=xxx API_URL2=xxx API_KEY=xxx eas update --branch preview`

**Reply by AdamGerthel** [reply]
I'm getting the exact same error, but I don't understand why the env variables would be empty. According to the console, they're being loaded:

<img width="551" alt="Screenshot 2024-02-21 at 10 12 14" src="https://github.com/expo/expo/assets/1490370/3878773a-69a0-4f86-9973-b92a12751973">
